protocols/gossipsub/src/handler.rs

```diff
@@ -338,28 +295,17 @@ impl ConnectionHandler for Handler {
             }
         }
 
-        if self.inbound_substreams_created > MAX_SUBSTREAM_CREATION {
-            // Too many inbound substreams have been created, end the connection.
-            return Poll::Ready(ConnectionHandlerEvent::Close(
-                HandlerError::MaxInboundSubstreams,
-            ));
-        }
+        // Invariant: `self.inbound_substreams_created < MAX_SUBSTREAM_CREATION`.
```

Comment:
How about adding an `assert!` or `debug_assert!` (preference for the former) here?

Reply:
Taking a closer look, I don't think this invariant actually holds.

In `on_connection_event` we drop new inbound requests on `==`:

``` rust
        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
            // Too many inbound substreams have been created, disable the handler.
            self.keep_alive = KeepAlive::No;
            log::warn!("The maximum number of inbound substreams created has been exceeded");
            return;
        }
```

Thus `self.inbound_substreams` may never be `>` but it could be `==` and thus the invariant of `<` does not hold.

@thomaseizinger am I missing something?